com.unity.netcode.gameobjects/Runtime/Serialization/BitReader.cs

```diff
@@ -0,0 +1,217 @@
+                {
+                    ReadMisaligned(out asBytes[i]);
+                }
+            }
+
+            val |= (ulong)ReadByteBits((int)bitCount & 7) << ((int)bitCount & ~7);
+            value = val;
+        }
+
+        /// <summary>
+        /// Read bits from stream.
+        /// </summary>
+        /// <param name="value">Value to store bits into.</param>
+        /// <param name="bitCount">Amount of bits to read.</param>
+        public void ReadBits(out byte value, uint bitCount)
+        {
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+            int checkPos = (int)(m_BitPosition + bitCount);
+            if (checkPos > m_AllowedBitwiseReadMark)
+            {
+                throw new OverflowException("Attempted to read without first calling TryBeginReadBits()");
+            }
+#endif
+            value = ReadByteBits((int)bitCount);
+        }
+
+        /// <summary>
+        /// Read a single bit from the buffer
+        /// </summary>
+        /// <param name="bit">Out value of the bit. True represents 1, False represents 0</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe void ReadBit(out bool bit)
+        {
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+            int checkPos = (m_BitPosition + 1);
+            if (checkPos > m_AllowedBitwiseReadMark)
+            {
+                throw new OverflowException("Attempted to read without first calling TryBeginReadBits()");
+            }
+#endif
+
+            int offset = m_BitPosition & 7;
+            int pos = m_BitPosition >> 3;
+            bit = (m_BufferPointer[pos] & (1 << offset)) != 0;
+            ++m_BitPosition;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe void ReadPartialValue<T>(out T value, int bytesToRead, int offsetBytes = 0) where T : unmanaged
+        {
+            var val = new T();
+            byte* ptr = ((byte*)&val) + offsetBytes;
+            byte* bufferPointer = m_BufferPointer + m_Position;
+            BytewiseUtility.FastCopyBytes(ptr, bufferPointer, bytesToRead);
+
+            m_BitPosition += bytesToRead * k_BitsPerByte;
+            value = val;
+        }
+
+        private byte ReadByteBits(int bitCount)
+        {
+            if (bitCount > 8)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bitCount), "Cannot read more than 8 bits into an 8-bit value!");
+            }
+
+            if (bitCount < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bitCount), "Cannot read fewer than 0 bits!");
+            }
+
+            int result = 0;
+            var convert = new ByteBool();
+            for (int i = 0; i < bitCount; ++i)
+            {
+                ReadBit(out bool bit);
+                result |= convert.Collapse(bit) << i;
+            }
+
+            return (byte)result;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
I'm not entirely sure how much we gain from aggressive inlining attributes like these — do you have any metrics or comparison?

Reply:
It wasn't earth-shattering but it was significant. Like 20-30% gain IIRC. But I'll rerun those tests tomorrow.

Reply:
> 20-30%

:O

Reply:
Here's the tests. Tested both aggressive inlining and branching.

Editor:
```
Inlined/No Branch: 18,707,485 ticks
Not Inlined/No Branch: 19,292,008 ticks (+3.12%)
Inlined/With Branch: 24,172,988 ticks (+29.21%)
Not Inlined/With Branch: 24,214,432 ticks (+29.43%)
```
Player:
```
Inlined/No Branch: 3,940,828 ticks
Not Inlined/No Branch: 5,748,679 ticks (+45.87%)
Inlined/With Branch: 4,384,911 ticks (+11.2%)
Not Inlined/With Branch: 7,607,113 ticks (+93.03%)
```
IL2CPP:
```
Inlined/No Branch: 2,144,231 ticks
Not Inlined/No Branch: 3,316,924 ticks (+54.69%)
Inlined/With Branch: 3,770,607 ticks (+75.84%)
Not Inlined/With Branch: 3,719,728 ticks (73.47%)
```

Reply:
Just for the same of completeness can we add the code used to get these perf numbers into the test project ? Might just be a good reference.

Reply:
Where and how should this kind of code be added to the test project, given that it's not really an automated test?

Reply:
We have code in the test project that are manual tests... I mean its a basic unity project we have all sort of things in there its not a project restricted to just automated testing. We do also use it as the project for that when running package tests because it just makes the package ci process easier.